docs/design/Ratify v2 Architecture Proposal.md

```diff
@@ -0,0 +1,262 @@
+5. Even though the current executor has been optimized for performance, it will be a bottleneck when the number of security artifacts increased significantly.
+6. The configuration applied in CLI and K8s scenarios are not exactly same, which makes it difficult for users to maintain and switch between different scenarios.
+
+## Proposed Architecture of Ratify v2
+
+### High Level Design
+```mermaid
+    %% Subgraph for inputs
+    flowchart TD
+    subgraph Inputs [Inputs]
+        direction TB
+        C[User Input per Request]
+        D["System Input (config/CRD)"]
+    end
+
+    %% Subgraph for core logic
+    subgraph Core [Core Logic]
+        E[core library]
+    end
+    
+    %% Subgraph for middleware
+    subgraph Middleware [Middlewares]
+        direction TB
+        B[Driver/Entrypoint] --> F[Output Render]
+    end
+    
+    A[Customized App]
+
+    %% Relationships
+    Inputs --> B
+    B --> Core
+    A -.- B
+    A --> Core
+```
+
+#### Architecture Components
+1. **Core Logic**: The core logic of Ratify v2 will be implemented as a separate library, which can be used by different scenarios. The core logic will be responsible for the validation of security metadata of any given artifacts. It will be designed to be extensible so that it can be easily integrated with different scenarios. And it will be optimized for performance and scalability.
+2. **User Input per Request**: The user input is the validation request from the user to validate the security metadata of the artifacts. The user input will be handled by the driver/entrypoint and passed to the core logic for validation.
+3. **System Input (config/CRD)**: The system input defines configurations of the Ratify application. It can be a configuration file or a custom resource definition (CRD) in Kubernetes. The system input can be configured during startup or runtime.
+4. **Driver/Entrypoint**: The driver/entrypoint will be responsible for handling user input and system input, and invoking the core logic to validate the security metadata of the artifacts. It will also be responsible for invoking the output render to render the validation results. It will have different implementations in terms of the user scenario.
+5. **Output Render**: The output render will be responsible for rendering the validation results.
+6. **Middlewares**: The middleware consists of `Driver/Entrypoint` and `Output Render`. Each user scenario will have its own middleware implementation.
+7. **Customized App**: The customized app can be any application behaves in a similar was as the middleware. It will be responsible for handling its own input and invoking the core logic to validate the security metadata of the artifacts and rendering its own output.
+
+### Deisgn Details
+Below is a more detailed design of the proposed architecture of Ratify v2. Note that text in red indicates a repository under `ratify-project` organization.
```

Comment:
the v2 diagram doesn't show the cache component. Will all plugins share the oras cache in v2?

Reply:
yes, the new diagram focuses on the core and plugins. Oras cache will reside in the oras implementation. And since all plugins will run in the host process as library, the oras cache is also in the host process which can be safely shared by verifiers. I'll add it to the doc as well.